encoding/codecv4/codecv4.go

```diff
@@ -0,0 +1,446 @@
+package codecv4
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/crypto"
+	"github.com/scroll-tech/go-ethereum/crypto/kzg4844"
+	"github.com/scroll-tech/go-ethereum/log"
+
+	"github.com/scroll-tech/da-codec/encoding"
+	"github.com/scroll-tech/da-codec/encoding/codecv3"
+	"github.com/scroll-tech/da-codec/encoding/zstd"
+)
+
+// MaxNumChunks is the maximum number of chunks that a batch can contain.
+const MaxNumChunks = codecv3.MaxNumChunks
+
+// DABlock represents a Data Availability Block.
+type DABlock = codecv3.DABlock
+
+// DAChunk groups consecutive DABlocks with their transactions.
+type DAChunk = codecv3.DAChunk
+
+// DABatch contains metadata about a batch of DAChunks.
+type DABatch struct {
+	// header
+	Version              uint8          `json:"version"`
+	BatchIndex           uint64         `json:"batch_index"`
+	L1MessagePopped      uint64         `json:"l1_message_popped"`
+	TotalL1MessagePopped uint64         `json:"total_l1_message_popped"`
+	DataHash             common.Hash    `json:"data_hash"`
+	BlobVersionedHash    common.Hash    `json:"blob_versioned_hash"`
+	ParentBatchHash      common.Hash    `json:"parent_batch_hash"`
+	LastBlockTimestamp   uint64         `json:"last_block_timestamp"`
+	BlobDataProof        [2]common.Hash `json:"blob_data_proof"`
+
+	// blob payload
+	blob *kzg4844.Blob
+	z    *kzg4844.Point
+
+	// for batch task
+	blobBytes []byte
+}
+
+// NewDABlock creates a new DABlock from the given encoding.Block and the total number of L1 messages popped before.
+func NewDABlock(block *encoding.Block, totalL1MessagePoppedBefore uint64) (*DABlock, error) {
+	return codecv3.NewDABlock(block, totalL1MessagePoppedBefore)
+}
+
+// NewDAChunk creates a new DAChunk from the given encoding.Chunk and the total number of L1 messages popped before.
+func NewDAChunk(chunk *encoding.Chunk, totalL1MessagePoppedBefore uint64) (*DAChunk, error) {
+	return codecv3.NewDAChunk(chunk, totalL1MessagePoppedBefore)
+}
+
+// NewDABatch creates a DABatch from the provided encoding.Batch.
+func NewDABatch(batch *encoding.Batch, enableCompress bool) (*DABatch, error) {
```

Comment:
How does the caller set `enableCompress`? Try with `true` first and if it errors then try again with `false`?

Reply:
Yes.
in l2geth-verifier: try with `true` first and if it errors then try again with `false`.
in rollup-relayer: will ask `CheckChunkCompressedDataCompatibility` and `CheckBatchCompressedDataCompatibility`, with some strategies to determine whether or not to use `enableCompress`.